src/tree/gpu_hist/histogram.cu

```diff
@@ -169,33 +181,41 @@ void BuildGradientHistogram(EllpackDeviceAccessor const& matrix,
 
   // determine the launch configuration
   unsigned block_threads = shared ? 1024 : 256;
+  int num_groups = feature_groups.size() - 1;
   int n_mps = 0;
   dh::safe_cuda(cudaDeviceGetAttribute(&n_mps, cudaDevAttrMultiProcessorCount, device));
   int n_blocks_per_mp = 0;
   dh::safe_cuda(cudaOccupancyMaxActiveBlocksPerMultiprocessor
                 (&n_blocks_per_mp, kernel, block_threads, smem_size));
+  // TODO(canonizer): This is really a hack, find a better way to distribute the data
+  // among thread blocks
   unsigned grid_size = n_blocks_per_mp * n_mps;
+  grid_size = common::DivRoundUp(grid_size, common::DivRoundUp(num_groups, 4));
```

Comment:
I don't understand num_groups/4.

Needs comments here explaining the work decomposition.

Reply:
Done.